modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/certificatevalidation/CertificateVerificationManager.java

```diff
@@ -0,0 +1,282 @@
+
+                try {
+                    aliases = trustStore.aliases();
+                } catch (KeyStoreException e) {
+                    throw new CertificateVerificationException("Error while retrieving aliases from truststore", e);
+                }
+
+                while (aliases.hasMoreElements()) {
+
+                    alias = aliases.nextElement();
+                    try {
+                        issuerCert = (X509Certificate) trustStore.getCertificate(alias);
+                    } catch (KeyStoreException e) {
+                        throw new CertificateVerificationException("Unable to read the certificate from " +
+                                "truststore with the alias: " + alias, e);
+                    }
+
+                    if (issuerCert == null) {
+                        throw new CertificateVerificationException("Issuer certificate not found in truststore");
+                    }
+
+                    try {
+                        peerCert.verify(issuerCert.getPublicKey());
+
+                        log.debug("Valid issuer certificate found in the client truststore. Caching..");
+
+                        // Store the valid issuer cert in cache for future use
+                        certCache.setCacheValue(peerCert.getSerialNumber().toString(), issuerCert);
+                        if (log.isDebugEnabled()) {
+                            log.debug("Issuer certificate with serial number: " + issuerCert.getSerialNumber()
+                                    .toString() + " has been cached against the serial number:  " + peerCert
+                                    .getSerialNumber().toString() + " of the peer certificate.");
+                        }
+                        break;
+                    } catch (SignatureException | CertificateException | NoSuchAlgorithmException |
+                             InvalidKeyException | NoSuchProviderException e) {
+                        // Unable to verify the signature. Check with the next certificate in the next loop traversal.
+                    }
+                }
+            } else {
+                X509Certificate cachedIssuerCert = certCache.getCacheValue(peerCert.getSerialNumber().toString());
+                try {
+                    peerCert.verify(cachedIssuerCert.getPublicKey());
+                } catch (SignatureException | CertificateException | NoSuchAlgorithmException |
+                         InvalidKeyException |
+                         NoSuchProviderException e) {
+                    // Unable to verify the signature.
+                    throw new CertificateVerificationException("Unable to verify the signature of the certificate.");
+                }
+            }
+        }
+
+        long start = System.currentTimeMillis();
+
+        OCSPCache ocspCache = OCSPCache.getCache();
+        ocspCache.init(cacheSize, cacheDelayMins);
```

Comment:
shouldn't we initialize caches at the init, rather than doing it for each and every validation?

Reply:
Fixed with https://github.com/wso2/wso2-synapse/pull/2119/commits/09b48b314fa878b08f5287ee7ab2026ac5f29a01